Dockerfile

```diff
@@ -8,54 +9,79 @@ RUN cd llama.cpp && \
     make && \
     mv main llama
 
-# Copy over rest of the project files
+# Base image for node
+FROM node:19 as node_base
 
-FROM ubuntu:22.04 as deployment
+WORKDIR /usr/src/app
+# Install pip and requirements
+
+# Base image for runtime
+FROM ubuntu:22.04 as base
 
 ARG DEBIAN_FRONTEND=noninteractive
 ENV TZ=Europe/Amsterdam
 
 WORKDIR /usr/src/app
 
+# Install MongoDB and necessary tools
+RUN apt update && \
+    apt install -y curl wget gnupg && \
+    wget -qO - https://www.mongodb.org/static/pgp/server-6.0.asc | apt-key add - && \
+    echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu jammy/mongodb-org/6.0 multiverse" | tee /etc/apt/sources.list.d/mongodb-org-6.0.list && \
+    apt-get update && \
+    apt-get install -y mongodb-org
+
+
 # install pip
-RUN apt update
-RUN apt-get install -y python3-pip curl wget
+RUN apt update && \
+    apt install -y python3-pip
 RUN pip install --upgrade pip
 
-# install nodejs
-RUN curl -sL https://deb.nodesource.com/setup_19.x | bash
-RUN apt-get install nodejs
-
+# copy & install python reqs
+COPY ./api/requirements.txt api/requirements.txt
+RUN pip install --upgrade pip && \
+    pip install --no-cache-dir -r ./api/requirements.txt
 
-# MongoDB
-RUN wget -qO - https://www.mongodb.org/static/pgp/server-6.0.asc | apt-key add -
-RUN echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu jammy/mongodb-org/6.0 multiverse" | tee /etc/apt/sources.list.d/mongodb-org-6.0.list
+# Copy files
+COPY --from=llama_builder /tmp/llama.cpp/llama /usr/bin/llama
 
+# Dev environment
+FROM base as dev
+ENV NODE_ENV='development'
 
-RUN apt-get update
-RUN apt-get install -y mongodb-org
+# Install Node.js and npm packages
+COPY --from=node_base /usr/local /usr/local
+COPY ./web/package*.json ./
+RUN npm ci
 
+# Copy the rest of the project files
+COPY web /usr/src/app/web
+COPY ./api /usr/src/app/api
 
-# install requirements
-COPY ./api/requirements.txt api/requirements.txt
-RUN pip install -r ./api/requirements.txt
+COPY dev.sh /usr/src/app/dev.sh
+RUN chmod +x /usr/src/app/dev.sh
+CMD ./dev.sh
 
-# copy llama binary from llama_builder
-COPY --from=llama_builder /tmp/llama.cpp/llama /usr/bin/llama
+# Build frontend
+FROM node_base as frontend_builder
 
-# copy built webserver from web_builder
-COPY ./web/package.json web/
-COPY ./web/package-lock.json web/
-# Install Node modules
-RUN cd web && npm install
+COPY ./web/package*.json ./
+RUN npm ci
 
 COPY ./web /usr/src/app/web/
+WORKDIR /usr/src/app/web/
+RUN npm run build
+
+# Runtime environment
+FROM base as release
+
+ENV NODE_ENV='production'
```

Comment:
I dont think this line is needed. At this point Nodejs is not installed anymore

Reply:
We still use this variable for FastAPI setup. I figured we'd use the same env variable name

Reply:
Sounds good